resource/model.go

```diff
@@ -0,0 +1,140 @@
+	shortModelRegexValidator = regexp.MustCompile(`^([\w-]+)$`)
+)
+
+// ModelFamily is a family of related models.
+type ModelFamily struct {
+	Namespace   Namespace       `json:"namespace"`
+	ModelFamily ModelFamilyName `json:"model_family"`
+}
+
+// NewModelFamily creates a new ModelFamily based on parameters passed in.
+func NewModelFamily(namespace Namespace, family ModelFamilyName) ModelFamily {
+	return ModelFamily{namespace, family}
+}
+
+// Validate ensures that important fields exist and are valid.
+func (f ModelFamily) Validate() error {
+	if f.Namespace == "" {
+		return errors.New("model namespace field for resource missing")
+	}
+	if f.ModelFamily == "" {
+		return errors.New("model family field for resource missing")
+	}
+	if err := ContainsReservedCharacter(string(f.Namespace)); err != nil {
+		return err
+	}
+	if err := ContainsReservedCharacter(string(f.ModelFamily)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// String returns the model family string for the resource.
+func (f ModelFamily) String() string {
+	return fmt.Sprintf("%s:%s", f.Namespace, f.ModelFamily)
+}
+
+// Model represents an individual model within a family.
+type Model struct {
+	ModelFamily `json:","`
+	Name        ModelName `json:"name"`
+}
+
+// NewModel creates a new Model based on parameters passed in.
+func NewModel(namespace Namespace, fName ModelFamilyName, model ModelName) Model {
+	family := NewModelFamily(namespace, fName)
+	return Model{family, model}
+}
+
+// NewDefaultModel creates a new Model in the rdk:builtin namespace/family based on parameters passed in.
+func NewDefaultModel(model ModelName) Model {
+	return Model{ModelFamilyDefault, model}
+}
+
+// NewModelFromString creates a new Name based on a fully qualified resource name string passed in.
+func NewModelFromString(modelStr string) (Model, error) {
+	if modelRegexValidator.MatchString(modelStr) {
+		matches := modelRegexValidator.FindStringSubmatch(modelStr)
+		return NewModel(Namespace(matches[1]), ModelFamilyName(matches[2]), ModelName(matches[3])), nil
+	}
+	if shortModelRegexValidator.MatchString(modelStr) {
+		return NewModel(ResourceNamespaceRDK, ModelFamilyDefaultName, ModelName(modelStr)), nil
+	}
+	return Model{}, errors.Errorf("string %q is not a valid model name", modelStr)
+}
+
+// Validate ensures that important fields exist and are valid.
+func (m Model) Validate() error {
+	if err := m.ModelFamily.Validate(); err != nil {
+		return err
+	}
+	if m.Name == "" {
+		return errors.New("model name field for resource missing")
+	}
+	if err := ContainsReservedCharacter(string(m.Name)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// String returns the resource model string for the component.
+func (m Model) String() string {
+	return fmt.Sprintf("%s:%s", m.ModelFamily, m.Name)
+}
+
+// UnmarshalJSON pareses namespace:family:modelname strings to the full Model{} struct.
+func (m *Model) UnmarshalJSON(data []byte) error {
```

Comment:
While the new "Model" type is structured in a way just like Subtype is, this makes it easy to simply define the model in one human readable string instead of a nested object. I find it easier/simpler to read, and think we should default to that for examples/documentation.